File: backend/src/controllers/appointment.controller.ts
import { Request, Response, NextFunction } from 'express';
import * as appointmentService from '../services/appointment.service';

export const bookAppointment = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const result = await appointmentService.bookAppointment(req.body);
    res.status(201).json({
      success: true,
      message: 'Appointment booked successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getAppointment = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await appointmentService.getAppointmentById(id);
    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getAllAppointments = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const page = parseInt(req.query.page as string || '0', 10);
    const size = parseInt(req.query.size as string || '20', 10);
    const status = req.query.status as string;

    const result = await appointmentService.getAllAppointments(page, size, status);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getPatientAppointments = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const patientId = parseInt(req.params.patientId, 10);
    const page = parseInt(req.query.page as string || '0', 10);
    const size = parseInt(req.query.size as string || '20', 10);

    const result = await appointmentService.getPatientAppointments(patientId, page, size);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getDoctorAppointments = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const doctorId = parseInt(req.params.doctorId, 10);
    const page = parseInt(req.query.page as string || '0', 10);
    const size = parseInt(req.query.size as string || '20', 10);

    const result = await appointmentService.getDoctorAppointments(doctorId, page, size);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const getTodayAppointments = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const doctorId = parseInt(req.params.doctorId, 10);
    const result = await appointmentService.getTodayAppointments(doctorId);

    res.status(200).json({
      success: true,
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const rescheduleAppointment = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const { newDateTime } = req.body;

    if (!newDateTime) {
      return res.status(400).json({
        success: false,
        message: 'Request body must include "newDateTime"',
      });
    }

    const result = await appointmentService.rescheduleAppointment(id, newDateTime);

    res.status(200).json({
      success: true,
      message: 'Appointment rescheduled successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const cancelAppointment = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const { reason } = req.body || {};

    const result = await appointmentService.cancelAppointment(id, reason);

    res.status(200).json({
      success: true,
      message: 'Appointment cancelled successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const completeAppointment = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const { consultationNotes } = req.body || {};

    const result = await appointmentService.completeAppointment(id, consultationNotes);

    res.status(200).json({
      success: true,
      message: 'Appointment completed successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

export const confirmAppointment = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id, 10);
    const result = await appointmentService.confirmAppointment(id);

    res.status(200).json({
      success: true,
      message: 'Appointment confirmed successfully',
      data: result,
    });
  } catch (error) {
    next(error);
  }
};

File: backend/hospital-backend/src/main/java/com/hospital/appointment/controller/AppointmentController.java
package com.hospital.appointment.controller;

import com.hospital.appointment.dto.*;
import com.hospital.appointment.service.AppointmentService;
import com.hospital.common.dto.ApiResponse;
import com.hospital.common.dto.PagedResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@RestController @RequestMapping("/appointments") @RequiredArgsConstructor
@Tag(name = "Appointments", description = "Appointment booking and management")
public class AppointmentController {

    private final AppointmentService appointmentService;

    @PostMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST', 'PATIENT', 'DOCTOR')")
    @Operation(summary = "Book Appointment")
    public ResponseEntity<ApiResponse<AppointmentResponse>> bookAppointment(
            @Valid @RequestBody BookAppointmentRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Appointment booked", appointmentService.bookAppointment(request)));
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get Appointment")
    public ResponseEntity<ApiResponse<AppointmentResponse>> getAppointment(@PathVariable Long id) {
        return ResponseEntity.ok(ApiResponse.success(appointmentService.getAppointmentById(id)));
    }

    @GetMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST')")
    @Operation(summary = "List All Appointments")
    public ResponseEntity<ApiResponse<PagedResponse<AppointmentResponse>>> getAllAppointments(
            @RequestParam(defaultValue = "0") int page, @RequestParam(defaultValue = "20") int size,
            @RequestParam(required = false) String status) {
        return ResponseEntity.ok(ApiResponse.success(appointmentService.getAllAppointments(page, size, status)));
    }

    @GetMapping("/patient/{patientId}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'RECEPTIONIST', 'PATIENT')")
    @Operation(summary = "Get Patient Appointments")
    public ResponseEntity<ApiResponse<PagedResponse<AppointmentResponse>>> getPatientAppointments(
            @PathVariable Long patientId, @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        return ResponseEntity.ok(ApiResponse.success(appointmentService.getPatientAppointments(patientId, page, size)));
    }

    @GetMapping("/doctor/{doctorId}")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR', 'RECEPTIONIST')")
    @Operation(summary = "Get Doctor Appointments")
    public ResponseEntity<ApiResponse<PagedResponse<AppointmentResponse>>> getDoctorAppointments(
            @PathVariable Long doctorId, @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        return ResponseEntity.ok(ApiResponse.success(appointmentService.getDoctorAppointments(doctorId, page, size)));
    }

    @GetMapping("/doctor/{doctorId}/today")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR')")
    @Operation(summary = "Get Today's Appointments for Doctor")
    public ResponseEntity<ApiResponse<List<AppointmentResponse>>> getTodayAppointments(@PathVariable Long doctorId) {
        return ResponseEntity.ok(ApiResponse.success(appointmentService.getTodayAppointments(doctorId)));
    }

    @PutMapping("/{id}/reschedule")
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST', 'PATIENT')")
    @Operation(summary = "Reschedule Appointment")
    public ResponseEntity<ApiResponse<AppointmentResponse>> rescheduleAppointment(
            @PathVariable Long id, @RequestBody Map<String, LocalDateTime> body) {
        return ResponseEntity.ok(ApiResponse.success("Appointment rescheduled",
                appointmentService.rescheduleAppointment(id, body.get("newDateTime"))));
    }

    @PutMapping("/{id}/cancel")
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST', 'PATIENT', 'DOCTOR')")
    @Operation(summary = "Cancel Appointment")
    public ResponseEntity<ApiResponse<AppointmentResponse>> cancelAppointment(
            @PathVariable Long id, @RequestBody(required = false) Map<String, String> body) {
        String reason = body != null ? body.get("reason") : null;
        return ResponseEntity.ok(ApiResponse.success("Appointment cancelled",
                appointmentService.cancelAppointment(id, reason)));
    }

    @PutMapping("/{id}/complete")
    @PreAuthorize("hasAnyRole('ADMIN', 'DOCTOR')")
    @Operation(summary = "Complete Appointment")
    public ResponseEntity<ApiResponse<AppointmentResponse>> completeAppointment(
            @PathVariable Long id, @RequestBody(required = false) Map<String, String> body) {
        String notes = body != null ? body.get("consultationNotes") : null;
        return ResponseEntity.ok(ApiResponse.success("Appointment completed",
                appointmentService.completeAppointment(id, notes)));
    }

    @PutMapping("/{id}/confirm")
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST', 'DOCTOR')")
    @Operation(summary = "Confirm Appointment")
    public ResponseEntity<ApiResponse<AppointmentResponse>> confirmAppointment(@PathVariable Long id) {
        return ResponseEntity.ok(ApiResponse.success("Appointment confirmed",
                appointmentService.confirmAppointment(id)));
    }
}
